### DISCUSSION TITLE
브라우저 렌더링

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#13] [프론트 기술 질문 공간] [answered]
브라우저 렌더링에 대해 아는대로 설명 해주세요.

**Comment by michoball** [comment]
브라우저에 화면이 렌더링 될 때는  

Parsing - Style - Layout - Paint - Composite 의 과정을 거칩니다.

Parsing  단계에서 서버로부터 받아온 HTML CSS를 해석해서 가각 dom tree와 cssom tree를 생성합니다. 

Style 단계에서 dom tree 와  sccom tree 를 합쳐서 render tree를 생성합니다.  여기서 실제 화면에 그려질 애들만 합쳐집니다. 

Layout 단계에서는 render tree를 화면에 어떤 위치에 어떤 크기로 배치할지 계산을 합니다. 

Paint 단계에서는 Layout에서 계산한 값을 바탕으로  render tree의 요소들을 화면에 배치를 합니다. 

Composite 단계에서  Paint에서 만든 레이어를 합성하여 실제 화면에 나타나게 합니다.

**Comment by WooGie911** [comment, marked as answer]
브라우저 렌더링 과정은 크게 5가지 과정으로 구분 할 수 있습니다. 

parsing -> style -> layout -> paint -> composite

먼저 parsing 단계입니다. 
브라우저는 html과 css 파일을 받아서 각각 파싱합니다. 
html 파일은 dom 트리가 되고, css 파일은 cssom 트리가 됩니다. 

두번째로 style 단계입니다. 
두 트리를 비교하며 결합하여 render tree를 만들게 됩니다. 
render tree에 구성된대로 화면에 보여지게 되며 display: none 속성을 갖는 경우 render tree에서 제외됩니다. 

세번째로 layout 단계입니다. 
layout 단계에서는 render tree를 어떻게 화면에 배치할지 노드의 정확한 위치와 크기를 계산합니다. 
만약 크기를 %단위로 지정하였다면, layout 단계에서 %값을 계산하여 픽셀로 변환합니다.

네번째로 paint 단계입니다. 
layout 단계에서 계산된 값을 이용하여 render tree의 각 노드를 실제 픽셀로 변환해줍니다. 
이때 픽셀로 변환된 결과는 하나의 레이어가 아니라 여러 개의 레이어로 관리됩니다. 

마지막으로 composite 단계입니다. 
Paint 단계에서 생성된 레이어를 합성하여 실제 화면에 나타냅니다. 
이를 통해 우리는 화면에서 웹 페이지를 볼 수 있습니다.

**Comment by skyhanull** [comment]
1. HTML를 파싱 후, DOM트리를 구축합니다.
2. CSS를 파싱 후, CSSOM트리를 구축합니다.
3. Javascript를 실행합니다.
    - 주의! HTML 중간에 스크립트가 있다면 HTML 파싱이 중단됩니다.
4. DOM과 CSSOM을 조합하여 랜더트리를 구축합니다.
    - 주의! `display: none` 속성과 같이 화면에서 보이지도 않고 공간을 차지하지 않는 것은 렌더트리로 구축되지 않습니다.
5. 뷰포트 기반으로 렌더트리의 각 노드가 가지는 정확한 위치와 크기를 계산합니다.
6. 계산한 위치/크기를 기반으로 화면에 그립니다.
